File: scripts/health_date_policy.py
from __future__ import annotations

import argparse
import json
from dataclasses import dataclass
from datetime import date, datetime, timedelta
from zoneinfo import ZoneInfo


IST = ZoneInfo("Asia/Kolkata")


@dataclass(frozen=True)
class HealthTargetContext:
    target_date: date
    policy: str
    label: str


def health_target_context(moment: datetime) -> HealthTargetContext:
    local = moment.astimezone(IST)
    if local.hour >= 20:
        return HealthTargetContext(local.date(), "D_EVENING", "D · evening cutoff")
    if local.hour < 2:
        return HealthTargetContext(local.date() - timedelta(days=1), "D_OVERNIGHT", "D · overnight window")
    return HealthTargetContext(local.date() - timedelta(days=1), "D_MINUS_1", "D-1")


def health_target_date(moment: datetime) -> date:
    return health_target_context(moment).target_date


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--at", help="ISO-8601 timestamp; defaults to now")
    parser.add_argument("--date-only", action="store_true")
    args = parser.parse_args()
    instant = datetime.fromisoformat(args.at) if args.at else datetime.now(IST)
    context = health_target_context(instant)
    if args.date_only:
        print(context.target_date.isoformat())
    else:
        print(json.dumps({
            "targetDate": context.target_date.isoformat(),
            "targetPolicy": context.policy,
            "targetLabel": context.label,
        }, ensure_ascii=False))

File: app/health-date-policy.ts
const IST_OFFSET_MS = 5.5 * 60 * 60 * 1000;

export type HealthTargetPolicy = "D_EVENING" | "D_OVERNIGHT" | "D_MINUS_1";

export type HealthTargetContext = {
  targetDate: string;
  policy: HealthTargetPolicy;
  label: string;
};

function shiftDateKey(dateKey: string, days: number) {
  const date = new Date(`${dateKey}T00:00:00Z`);
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

/**
 * Resolve the reporting date in Asia/Kolkata. The Health day advances at
 * 20:00 IST and the overnight D window stays anchored to the prior evening.
 */
export function healthTargetContext(at = new Date()): HealthTargetContext {
  const ist = new Date(at.getTime() + IST_OFFSET_MS);
  const calendarDate = ist.toISOString().slice(0, 10);
  const hour = ist.getUTCHours();

  if (hour >= 20) {
    return { targetDate: calendarDate, policy: "D_EVENING", label: "D · evening cutoff" };
  }
  if (hour < 2) {
    return { targetDate: shiftDateKey(calendarDate, -1), policy: "D_OVERNIGHT", label: "D · overnight window" };
  }
  return { targetDate: shiftDateKey(calendarDate, -1), policy: "D_MINUS_1", label: "D-1" };
}

export function healthTargetDateKey(at = new Date()) {
  return healthTargetContext(at).targetDate;
}

export function healthPolicyLabel(policy: HealthTargetPolicy | undefined) {
  if (policy === "D_EVENING") return "D · evening cutoff";
  if (policy === "D_OVERNIGHT") return "D · overnight window";
  return "D-1";
}
